pkg/cmd/config/options.go

```diff
@@ -0,0 +1,174 @@
+)
+
+// Options is a struct to support all config subcommands
+type options struct {
+	base.Options
+	// Configuration is the gardenctl configuration
+	Configuration *config.Config
+	// Command is the name of the config command
+	Command string
+	// Name is a unique name of this Garden that can be used to target this Garden
+	Name string
+	// KubeconfigFlag is the path to the kubeconfig file of the Garden cluster
+	KubeconfigFlag flag.StringFlag
+	// ContextFlag Overrides the current-context of the garden cluster kubeconfig
+	// +optional
+	ContextFlag flag.StringFlag
+	// Patterns is a list of regex patterns that can be defined to use custom input formats for targeting
+	// Use named capturing groups to match target values.
+	// Supported capturing groups: project, namespace, shoot
+	// +optional
+	Patterns []string
+}
+
+// Complete adapts from the command line args to the data required.
+func (o *options) Complete(f util.Factory, cmd *cobra.Command, args []string) error {
+	manager, err := f.Manager()
+	if err != nil {
+		return fmt.Errorf("failed to get target manager: %w", err)
+	}
+
+	config := manager.Configuration()
+	if config == nil {
+		return errors.New("failed to get configuration")
+	}
+
+	o.Configuration = config
+
+	switch o.Command {
+	case "set-garden", "delete-garden":
+		if len(args) > 0 {
+			o.Name = strings.TrimSpace(args[0])
+		}
+	case "view":
+		if o.Output == "" {
+			o.Output = "yaml"
+		}
+	}
+
+	return nil
+}
+
+// Validate validates the provided options
+func (o *options) Validate() error {
+	switch o.Command {
+	case "set-garden", "delete-garden":
+		if o.Name == "" {
+			return errors.New("garden identity is required")
+		}
+	case "view":
+		return o.Options.Validate()
+	}
+
+	return nil
+}
+
+// AddFlags adds flags to adjust the output to a cobra command
+func (o *options) AddFlags(flags *pflag.FlagSet) {
+	switch o.Command {
+	case "set-garden":
+		flags.Var(&o.KubeconfigFlag, "kubeconfig", "path to kubeconfig file for this Garden cluster")
+		flags.Var(&o.ContextFlag, "context", "override the current-context of the garden cluster kubeconfig")
+		flags.StringArrayVar(&o.Patterns, "pattern", nil, `define regex match patterns for this garden for custom input formats for targeting.
+Use named capturing groups to match target values.
+Supported capturing groups: project, namespace, shoot.
+Note that if you set this flag it will overwrite the pattern list in the config file.
+You may specify any number of extra patterns.
+Example: ^((?Pmy-garden[^/]+)/)?shoot--(?P<project>.+)--(?P<shoot>.+)$`)
+	case "delete-garden":
+		// no flags
+	case "view":
+		o.Options.AddFlags(flags)
+	}
+}
+
+// Run executes the command
+func (o *options) Run(_ util.Factory) error {
+	switch o.Command {
+	case "set-garden":
+		return o.runSetGarden()
+	case "delete-garden":
+		return o.runDeleteGarden()
+	case "view":
+		return o.runView()
+	}
+
+	return nil
+}
+
+func (o *options) runSetGarden() error {
```

Comment:
I understand why you moved the code here. As discussed we do not want to have the StringFlags in the config struct. However, what I do not like is the fact that for delete garden there is a function on the struct and for set garden all the logic is here. Maybe we could still have it in the struct. If we do not want to pass the StringFlag types we could also pass string pointers, where nil stands for value not provided. But that is just my personal opinion.

Reply:
If we don't that somebody operates on the struct data we have to hide it behind an interface. As of now we should consider the method of the Config struct as utility methods which could be used or not. To make it consistent i have moved the delete garden impl from the config struct to the command.